compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2553,4 +2565,150 @@ public int hashCode() {
+            case TypeTags.RECORD:
+                return checkFillerValue((BRecordType) type);
+            case TypeTags.TUPLE:
+                BTupleType tupleType = (BTupleType) type;
+                return tupleType.getTupleTypes().stream().allMatch(eleType -> hasFillerValue(eleType));
+            default:
+                return false;
+        }
+    }
+
+    private boolean checkFillerValue(BObjectType type) {
+        if ((type.tsymbol.flags & Flags.ABSTRACT) == Flags.ABSTRACT) {
+            return false;
+        }
+
+        BAttachedFunction initFunction = ((BObjectTypeSymbol) type.tsymbol).initializerFunc;
+        if (initFunction == null) {
+            return true;
+        }
+        if (initFunction.symbol.getReturnType().getKind() != TypeKind.NIL) {
+            return false;
+        }
+
+        for (BVarSymbol bVarSymbol : initFunction.symbol.getParameters()) {
+            if (!bVarSymbol.defaultableParam) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * This will handle two types. Singleton : As singleton can have one value that value should it self be a valid fill
+     * value Union : 1. if nil is a member it is the fill values 2. else all the values should belong to same type and
+     * the default value for that type should be a member of the union precondition : value space should have at least
+     * one element
+     *
+     * @param type BFiniteType union or finite
+     * @return boolean whether type has a valid filler value or not
+     */
+    private boolean checkFillerValue(BFiniteType type) {
+        if (type.isNullable()) {
+            return true;
+        }
+        if (type.getValueSpace().size() == 1) { // For singleton types, that value is the implicit initial value
+            return true;
+        }
+        Iterator iterator = type.getValueSpace().iterator();
+        BLangExpression firstElement = (BLangExpression) iterator.next();
+        boolean defaultFillValuePresent = isImplicitDefaultValue(firstElement);
+
+        while (iterator.hasNext()) {
+            BLangExpression value = (BLangExpression) iterator.next();
+            if (!isSameType(value.type, firstElement.type)) {
+                return false;
+            }
+            if (!defaultFillValuePresent && isImplicitDefaultValue(value)) {
+                defaultFillValuePresent = true;
+            }
+        }
+
+        return defaultFillValuePresent;
+    }
+
+    private boolean checkFillerValue(BUnionType type) {
+        if (type.isNullable()) {
+            return true;
+        }
+        Iterator<BType> iterator = type.getMemberTypes().iterator();
+        BType firstMember = iterator.next();
+        while (iterator.hasNext()) {
+            if (!isSameType(firstMember, iterator.next())) {
+                return false;
+            }
+        }
+        return isValueType(firstMember) && hasFillerValue(firstMember);
```

Comment:
The following should work right?

```ballerina
const FOO = "foo";
const BAR = "bar";

type FooBar FOO|BAR;

public function main() {
    (map<FooBar>|map<string>)[2] x = [];
}
```

I guess it won't work with the `isValueType` check?

Reply:
Shall we create an issue for this?

Reply:
Created https://github.com/ballerina-platform/ballerina-lang/issues/21251